# based on binutils 2.28
diff --git a/bfd/config.bfd b/bfd/config.bfd
index 1b28016..be156cb 100644
--- a/bfd/config.bfd
+++ b/bfd/config.bfd
@@ -816,6 +816,11 @@ case "${targ}" in
     targ_defvec=x86_64_elf64_vec
     want64=true
     ;;
+  x86_64-*-bubackos*)
+    targ_defvec=x86_64_elf64_vec
+    targ_selvecs=i386_elf32_vec
+    want64=true
+    ;;
 #endif
   i[3-7]86-*-lynxos*)
     targ_defvec=i386_elf32_vec
diff --git a/config.sub b/config.sub
index 70a4180..67a1228 100755
--- a/config.sub
+++ b/config.sub
@@ -1407,7 +1407,7 @@ case $os in
 	      | -morphos* | -superux* | -rtmk* | -rtmk-nova* | -windiss* \
 	      | -powermax* | -dnix* | -nx6 | -nx7 | -sei* | -dragonfly* \
 	      | -skyos* | -haiku* | -rdos* | -toppers* | -drops* | -es* \
-	      | -onefs* | -tirtos* | -phoenix* | -fuchsia*)
+	      | -onefs* | -tirtos* | -phoenix* | -fuchsia* | -bubackos*)
 	# Remember, each alternative MUST END IN *, to match a version number.
 		;;
 	-qnx*)
diff --git a/gas/configure.tgt b/gas/configure.tgt
index 711d537..6142a12 100644
--- a/gas/configure.tgt
+++ b/gas/configure.tgt
@@ -492,6 +492,7 @@ case ${generic_target} in
   *-*-aros*)				fmt=elf em=linux ;;
   *-*-vxworks* | *-*-windiss)		fmt=elf em=vxworks ;;
   *-*-netware)				fmt=elf em=netware ;;
+  i386-*-bubackos*)             fmt=elf em=gnu ;;
 esac
 
 case ${cpu_type} in
diff --git a/ld/Makefile.am b/ld/Makefile.am
index 15beaa7..2b4347b 100644
--- a/ld/Makefile.am
+++ b/ld/Makefile.am
@@ -295,6 +295,7 @@ ALL_EMULATION_SOURCES = \
 	eelf32xtensa.c \
 	eelf_i386.c \
 	eelf_i386_be.c \
+    eelf_i386_bubackos.c \
 	eelf_i386_chaos.c \
 	eelf_i386_fbsd.c \
 	eelf_i386_ldso.c \
@@ -511,6 +512,7 @@ ALL_64_EMULATION_SOURCES = \
 	eelf_k1om.c \
 	eelf_k1om_fbsd.c \
 	eelf_x86_64.c \
+    eelf_x86_64_bubackos.c \
 	eelf_x86_64_cloudabi.c \
 	eelf_x86_64_fbsd.c \
 	eelf_x86_64_nacl.c \
@@ -1333,6 +1335,10 @@ eelf32xtensa.c: $(srcdir)/emulparams/elf32xtensa.sh $(ELF_DEPS) \
 eelf_i386.c: $(srcdir)/emulparams/elf_i386.sh \
   $(ELF_X86_DEPS) $(srcdir)/scripttempl/elf.sc ${GEN_DEPENDS}
 
+eelf_i386_bubackos.c: $(srcdir)/emulparams/elf_i386_bubackos.sh \
+  $(ELF_DEPS) $(srcdir)/scripttempl/elf.sc ${GEN_DEPENDS}
+	${GENSCRIPTS} elf_i386_bubackos "$(tdir_elf_i386_bubackos)"
+
 eelf_i386_be.c: $(srcdir)/emulparams/elf_i386_be.sh \
   $(ELF_DEPS) $(srcdir)/scripttempl/elf.sc ${GEN_DEPENDS}
 
@@ -2032,6 +2038,10 @@ eelf_k1om_fbsd.c: $(srcdir)/emulparams/elf_k1om_fbsd.sh \
 eelf_x86_64.c: $(srcdir)/emulparams/elf_x86_64.sh \
   $(ELF_X86_DEPS) $(srcdir)/scripttempl/elf.sc ${GEN_DEPENDS}
 
+eelf_x86_64_bubackos.c: $(srcdir)/emulparams/elf_x86_64_bubackos.sh \
+  $(ELF_DEPS) $(srcdir)/scripttempl/elf.sc ${GEN_DEPENDS}
+	${GENSCRIPTS} elf_x86_64_bubackos "$(tdir_elf_x86_64_bubackos)"
+
 eelf_x86_64_cloudabi.c: $(srcdir)/emulparams/elf_x86_64_cloudabi.sh \
   $(srcdir)/emulparams/elf_x86_64.sh \
   $(ELF_X86_DEPS) $(srcdir)/scripttempl/elf.sc ${GEN_DEPENDS}
diff --git a/ld/configure.tgt b/ld/configure.tgt
index b85c6bb..a0998c4 100644
--- a/ld/configure.tgt
+++ b/ld/configure.tgt
@@ -245,6 +245,15 @@ i[3-7]86-*-aros*)	targ_emul=elf_i386
 			targ_extra_emuls=elf_iamcu ;;
 i[3-7]86-*-rdos*)	targ_emul=elf_i386
 			targ_extra_emuls=elf_iamcu ;;
+i[3-7]86-*-bubackos*)
+            targ_emul=elf_i386_bubackos
+            targ_extra_emuls=elf_i386
+            targ64_extra_emuls="elf_x86_64_bubackos elf_x86_64"
+            ;;
+x86_64-*-bubackos*)
+            targ_emul=elf_x86_64_bubackos
+            targ_extra_emuls="elf_i386_bubackos elf_x86_64 elf_i386"
+            ;;
 x86_64-*-rdos*)		targ_emul=elf64rdos ;;
 x86_64-*-cloudabi*)	targ_emul=elf_x86_64_cloudabi ;;
 i[3-7]86-*-bsd)		targ_emul=i386bsd ;;
diff --git a/ld/emulparams/elf_i386_bubackos.sh b/ld/emulparams/elf_i386_bubackos.sh
new file mode 100644
index 0000000..114b9c4
--- /dev/null
+++ b/ld/emulparams/elf_i386_bubackos.sh
@@ -0,0 +1,4 @@
+. ${srcdir}/emulparams/elf_i386.sh
+GENERATE_SHLIB_SCRIPT=yes
+GENERATE_PIE_SCRIPT=yes
+
diff --git a/ld/emulparams/elf_x86_64_bubackos.sh b/ld/emulparams/elf_x86_64_bubackos.sh
new file mode 100644
index 0000000..cafdb74
--- /dev/null
+++ b/ld/emulparams/elf_x86_64_bubackos.sh
@@ -0,0 +1,2 @@
+. ${srcdir}/emulparams/elf_x86_64.sh
+
